src/compiler/crystal/semantic/top_level_visitor.cr

```diff
@@ -949,9 +950,10 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
 
     node.body.accept self
 
-    # Now remove these vars, but only if they weren't vars before
-    node.args.each do |arg|
-      @vars.delete(arg.name) unless old_vars_keys.includes?(arg.name)
+    # After the block we should have the same number of local vars
+    # (blocks can't declare inject local vars to the outer scope)
+    while @vars.size > old_vars_size
```

Comment:
Would there be any difference if this was `(old_vars_size - @vars.size).times do` ? I doubt it is noticeable but the method is called a lot of times, so who knows? 

I wish we had an ordered map type, then we could have just nuked the last n entries in a more efficient manner.

EDIT: Hmm. Isn't the insertion order kept track of? Perhaps it is possible to do bulk operations. :thinking: 

/probably not worth the trouble though

Reply:
All excellent observations. Yeah, Hash is ordered and so is Set so we could nuke the last entries, though because it's a Hash we still need to find the key hash to see in which bucket it's located, clear it and so on, so probably the same performance in the end.